edx_event_bus_kafka/internal/consumer.py

```diff
@@ -85,45 +96,54 @@ def inner_from_dict(event_data_dict, ctx=None):  # pylint: disable=unused-argume
 
         return DeserializingConsumer(consumer_config)
 
+    def _shut_down(self):
+        """
+        Test utility for shutting down the consumer loop.
+        """
+        self._shut_down_loop = True
+
     def consume_indefinitely(self):
         """
         Consume events from a topic in an infinite loop.
         """
         # This is already checked at the Command level, but it's possible this loop
         # could get called some other way, so check it here too.
         if not KAFKA_CONSUMERS_ENABLED.is_enabled():
-            logger.error("Kafka consumers not enabled")
+            logger.error("Kafka consumers not enabled, exiting.")
             return
 
         try:
             full_topic = get_full_topic(self.topic)
+            run_context = {
+                'full_topic': full_topic,
+                'consumer_group': self.group_id,
+                'expected_signal': self.signal,
+            }
             self.consumer.subscribe([full_topic])
+            logger.info(
+                f"Running consumer for {run_context!r}; assigned {self.consumer.assignment()!r}"
+            )
 
-            # TODO: Make sure exceptions won't kill the loop. https://github.com/openedx/event-bus-kafka/issues/62
             while True:
-                msg = self.consumer.poll(timeout=CONSUMER_POLL_TIMEOUT)
-                if msg is not None:
-                    self.process_single_message(msg)
+                # Allow unit tests to break out of loop
+                if self._shut_down_loop:
+                    break
+
+                msg = None
+                try:
+                    msg = self.consumer.poll(timeout=CONSUMER_POLL_TIMEOUT)
+                    # DeserializingConsumer.poll() always returns either a valid message
+                    # or None, and raises an exception in all other cases. This means
+                    # we don't need to check msg.error() ourselves.
+                    if msg is not None:
+                        self.emit_signals_from_message(msg)
+                except BaseException as e:
+                    self.record_event_consuming_error(run_context, e, msg)
```

Comment:
[Author's note] I'm a little concerned that we could end up in a tight error-loop here under some conditions. May not be worth worrying about, though -- it would burn CPU, but not affect correctness.

Reply:
Ah yeah, it would be hell on the logs, and could be recording hundreds of stack traces per second. :-/ Maybe worth addressing now, then.

My instinct is this: If a message was received, don't worry about tight error loops. If we couldn't get a message _at all_, sleep for a short delay before continuing.

I think this would be relatively uncontroversial, but this approach is complicated by our use of DeserializingConsumer, whose poll call encompasses getting a message, talking to the schema registry, and actually deserializing. If we were using Consumer directly, we could separate out the error cases more easily because the poll call would be on another line, separated from deserialization.

DeserializingConsumer [documents](https://github.com/confluentinc/confluent-kafka-python/blob/395f93901d3296b89bd5b7f9c3de0a2a5a291411/src/confluent_kafka/deserializing_consumer.py#L95) three exceptions it may raise, but looking at the code it's non-exhaustive—I suspect that `Consumer.poll()` can actually raise exceptions in some cases. So perhaps the right thing to do is something like this pseudocode:

```
do_pause = False
try:
    poll()
except KeyDeserializationError, ValueDeserializationError:
    pass
except:
    do_pause = True  # should only be ConsumeError, but could be something unknown

...

# in finally: clause
if do_pause:
    time.sleep(1.0)
```

Reply:
We'll see if this is a bad call later, but I say we leave this as-is for now. Since we are committing, we should be down to one failure per message. We want to capture the failures in the log anyway. If it is noisy, it means there are lots of messages getting dropped, and hopefully this results in an alert and fixing the issue quickly.  Maybe this will force us to learn about specific errors that should be handled differently? Anyway, just a thought to keep the code clean and to move on and learn from reality.

Reply:
I believe we can currently have an arbitrarily large number of errors even with no messages -- it could raise a "couldn't talk to broker" error over and over again in a tight loop.

Reply:
Hmm, that's a thought. It would still sleep in cases where there was a schema mismatch, couldn't talk to schema registry, or bad Avro encoding; it would not sleep for unrecognized signal, other bad headers, or of course for receiver errors.

I could live with it. If we get into a situation where we're in a failure loop and sleeping on each iteration and not making progress, we could make the call to adjust the sleep length via remote config. (I'm not sure what exactly that situation might be—but the important thing is that we'd have an escape valve.) Would have to add it to the runbook, though.